lib/services/local/plugin_data.go

```diff
@@ -0,0 +1,219 @@
+				return trace.Wrap(err)
+			}
+			// In order to prevent orphaned plugin data, we automatically
+			// configure new instances to expire shortly after the resource
+			// to which they are associated.  This discrepency in expiry gives
+			// plugins the ability to use stored data when handling an expiry
+			// (OpDelete) event.
+			if parentExpiresAt == nil {
+				return trace.BadParameter("missing parent resource %q", params.Kind)
+			}
+			exp := *parentExpiresAt
+
+			data, err = types.NewPluginData(params.Resource, params.Kind)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			// A resource can have empty expiration time by default.
+			// In this case, we should leave PluginData expiration time empty as well.
+			// TODO: PluginData must be destroyed with parent resource. This needs to be addressed in future.
+			if !exp.IsZero() {
+				data.SetExpiry(exp.Add(time.Hour))
+			}
+			create = true
+		}
+		if err := data.Update(params); err != nil {
+			return trace.Wrap(err)
+		}
+		if err := data.CheckAndSetDefaults(); err != nil {
+			return trace.Wrap(err)
+		}
+		newItem, err := itemFromPluginData(data)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if create {
+			if _, err := s.Create(ctx, newItem); err != nil {
+				if trace.IsAlreadyExists(err) {
+					select {
+					case <-retry.After():
+						retry.Inc()
+						continue
+					case <-ctx.Done():
+						return trace.Wrap(ctx.Err())
+					}
+				}
+				return trace.Wrap(err)
+			}
+		} else {
+			if _, err := s.CompareAndSwap(ctx, *item, newItem); err != nil {
+				if trace.IsCompareFailed(err) {
+					select {
+					case <-retry.After():
+						retry.Inc()
+						continue
+					case <-ctx.Done():
+						return trace.Wrap(ctx.Err())
+					}
+				}
+				return trace.Wrap(err)
+			}
+		}
+		return nil
+	}
+	return trace.CompareFailed("too many concurrent writes to plugin data %s", params.Resource)
+}
+
+// pluginDataKey constructs PluginData entry backend key
+func pluginDataKey(kind string, name string) []byte {
+	return backend.Key(pluginDataPrefix, kind, name, paramsPrefix)
+}
+
+func itemFromPluginData(data types.PluginData) (backend.Item, error) {
+	value, err := services.MarshalPluginData(data)
+	if err != nil {
+		return backend.Item{}, trace.Wrap(err)
+	}
+	// enforce explicit limit on resource size in order to prevent PluginData from
+	// growing uncontrollably.
+	if len(value) > teleport.MaxResourceSize {
+		return backend.Item{}, trace.BadParameter("plugin data size limit exceeded")
+	}
+	return backend.Item{
+		Key:     pluginDataKey(data.GetSubKind(), data.GetName()),
+		Value:   value,
+		Expires: data.Expiry(),
+		ID:      data.GetResourceID(),
+	}, nil
+}
+
+func itemToPluginData(item backend.Item) (types.PluginData, error) {
+	data, err := services.UnmarshalPluginData(
+		item.Value,
+		services.WithResourceID(item.ID),
+		services.WithExpires(item.Expires),
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return data, nil
+}
+
```

Comment:
```suggestion
func isAllowedPluginDataKind(kind string) bool {
    return apiutils.SliceContainsStr([]string{types.KindAccessRequest, types.KindUser}, kind)
}
```

Reply:
In my opinion, checking slice worths it when a target list is dynamic. Here we check essentially a constant value, so, it seems that a simple if would be much readable and fast. Are you sure we need `SliceContainsStr` here?